src/hotspot/share/gc/g1/g1RemSet.cpp

```diff
@@ -1296,21 +1296,20 @@ class G1MergeHeapRootsTask : public WorkerTask {
       G1DirtyCardQueueSet& dcqs = G1BarrierSet::dirty_card_queue_set();
       BufferNodeList buffers = dcqs.take_all_completed_buffers();
 
-      size_t num_buffers = (buffers._entry_count / G1UpdateBufferSize);
-      size_t buffers_per_thread = MAX2(num_buffers / num_workers, (size_t)1);
+      size_t entries_per_thread = ceil(buffers._entry_count / (double)num_workers);
 
       BufferNode* head = buffers._head;
       BufferNode* tail = head;
 
       uint worker = 0;
       while (tail != nullptr) {
-        size_t count = 0;
+        size_t count = tail->size();
         BufferNode* cur = tail->next();
 
-        while (count < buffers_per_thread && cur != nullptr) {
+        while (count < entries_per_thread && cur != nullptr) {
           tail = cur;
+          count += tail->size();
           cur = tail->next();
-          count++;
         }
```

Comment:
Still good. Looking at this balancing code again, maybe a comment about that part ("// Balance dirty card buffers for better work distribution.") or factoring this out into a helper method for documentation purposes would be nice.
It just feels quite a bit of code in a constructor, but just a comment what this is doing is just fine. Or just ignore if you feel it is obvious.

Reply:
To me, the balancing code seems pretty obvious in the context of avoiding the bottleneck of a single
list being processed by multiple threads.  But a comment about that bottleneck to explain all of this stuff
around `_dirty_card_buffers` might be nice.

Reply:
Fwiw, I marked a random location in this code as basically the entire constructor is about balancing. Just commenting `_dirty_card_buffers` is just fine too (and in general, explaining _why_ something is done is preferable to _what_ is done which when describing the former typically contains the latter anyway).